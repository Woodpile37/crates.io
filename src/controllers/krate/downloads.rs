//! Endpoint for exposing crate download counts
//!
//! The endpoint for downloading a crate and exposing version specific
//! download counts are located in `version::downloads`.

use std::cmp;

use crate::controllers::frontend_prelude::*;

use crate::models::{Crate, CrateVersions, Version, VersionDownload};
use crate::schema::version_downloads;
use crate::sql::to_char;
use crate::util::errors::crate_not_found;
use crate::views::EncodableVersionDownload;

/// Handles the `GET /crates/:crate_id/downloads` route.
pub async fn downloads(state: AppState, Path(crate_name): Path<String>) -> AppResult<Json<Value>> {
    spawn_blocking(move || {
        use diesel::dsl::*;
        use diesel::sql_types::BigInt;

        let conn = &mut *state.db_read()?;
        let krate: Crate = Crate::by_name(&crate_name)
            .first(conn)
            .optional()?
            .ok_or_else(|| crate_not_found(&crate_name))?;

        let mut versions: Vec<Version> = krate.all_versions().load(conn)?;
        versions
            .sort_by_cached_key(|version| cmp::Reverse(semver::Version::parse(&version.num).ok()));
        let (latest_five, rest) = versions.split_at(cmp::min(5, versions.len()));

        let downloads = VersionDownload::belonging_to(latest_five)
            .filter(version_downloads::date.gt(date(now - 90.days())))
            .order((
                version_downloads::date.asc(),
                version_downloads::version_id.desc(),
            ))
            .load(conn)?
            .into_iter()
            .map(VersionDownload::into)
            .collect::<Vec<EncodableVersionDownload>>();

        let sum_downloads = sql::<BigInt>("SUM(version_downloads.downloads)");
        let extra: Vec<ExtraDownload> = VersionDownload::belonging_to(rest)
            .select((
                to_char(version_downloads::date, "YYYY-MM-DD"),
                sum_downloads,
            ))
            .filter(version_downloads::date.gt(date(now - 90.days())))
            .group_by(version_downloads::date)
            .order(version_downloads::date.asc())
            .load(conn)?;

        #[derive(Serialize, Queryable)]
        struct ExtraDownload {
            date: String,
            downloads: i64,
        }

        Ok(Json(json!({
            "version_downloads": downloads,
            "meta": {
                "extra_downloads": extra,
            },
        })))
    })
    .await
}
