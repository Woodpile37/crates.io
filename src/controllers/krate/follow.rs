//! Endpoints for managing a per user list of followed crates

use crate::auth::AuthCheck;
use diesel::associations::Identifiable;

use crate::controllers::frontend_prelude::*;
use crate::models::{Crate, Follow};
use crate::schema::*;
use crate::util::errors::crate_not_found;

fn follow_target(crate_name: &str, conn: &mut PgConnection, user_id: i32) -> AppResult<Follow> {
    let crate_id = Crate::by_name(crate_name)
        .select(crates::id)
        .first(conn)
        .optional()?
        .ok_or_else(|| crate_not_found(crate_name))?;

    Ok(Follow { user_id, crate_id })
}

/// Handles the `PUT /crates/:crate_id/follow` route.
pub async fn follow(
    app: AppState,
    Path(crate_name): Path<String>,
    req: Parts,
) -> AppResult<Response> {
    spawn_blocking(move || {
        let conn = &mut *app.db_write()?;
        let user_id = AuthCheck::default().check(&req, conn)?.user_id();
        let follow = follow_target(&crate_name, conn, user_id)?;
        diesel::insert_into(follows::table)
            .values(&follow)
            .on_conflict_do_nothing()
            .execute(conn)?;

        ok_true()
    })
    .await
}

/// Handles the `DELETE /crates/:crate_id/follow` route.
pub async fn unfollow(
    app: AppState,
    Path(crate_name): Path<String>,
    req: Parts,
) -> AppResult<Response> {
    spawn_blocking(move || {
        let conn = &mut *app.db_write()?;
        let user_id = AuthCheck::default().check(&req, conn)?.user_id();
        let follow = follow_target(&crate_name, conn, user_id)?;
        diesel::delete(&follow).execute(conn)?;

        ok_true()
    })
    .await
}

/// Handles the `GET /crates/:crate_id/following` route.
pub async fn following(
    app: AppState,
    Path(crate_name): Path<String>,
    req: Parts,
) -> AppResult<Json<Value>> {
    spawn_blocking(move || {
        use diesel::dsl::exists;

        let conn = &mut *app.db_read_prefer_primary()?;
        let user_id = AuthCheck::only_cookie().check(&req, conn)?.user_id();
        let follow = follow_target(&crate_name, conn, user_id)?;
        let following =
            diesel::select(exists(follows::table.find(follow.id()))).get_result::<bool>(conn)?;

        Ok(Json(json!({ "following": following })))
    })
    .await
}
