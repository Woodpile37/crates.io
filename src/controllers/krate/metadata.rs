//! Endpoints that expose metadata about a crate
//!
//! These endpoints provide data that could be obtained directly from the
//! index or cached metadata which was extracted (client side) from the
//! `Cargo.toml` file.

use std::cmp::Reverse;
use std::str::FromStr;

use crate::controllers::frontend_prelude::*;
use crate::controllers::helpers::pagination::PaginationOptions;

use crate::models::{
    Category, Crate, CrateCategory, CrateKeyword, CrateVersions, Keyword, RecentCrateDownloads,
    TopVersions, User, Version, VersionOwnerAction,
};
use crate::schema::*;
use crate::util::errors::crate_not_found;
use crate::views::{
    EncodableCategory, EncodableCrate, EncodableDependency, EncodableKeyword, EncodableVersion,
};

/// Handles the `GET /summary` route.
pub async fn summary(state: AppState) -> AppResult<Json<Value>> {
    spawn_blocking(move || {
        let config = &state.config;

        let conn = &mut *state.db_read()?;
        let num_crates: i64 = crates::table.count().get_result(conn)?;
        let num_downloads: i64 = metadata::table
            .select(metadata::total_downloads)
            .get_result(conn)?;

        fn encode_crates(
            conn: &mut PgConnection,
            data: Vec<(Crate, Option<i64>)>,
        ) -> AppResult<Vec<EncodableCrate>> {
            let recent_downloads = data.iter().map(|&(_, s)| s).collect::<Vec<_>>();

            let krates = data.into_iter().map(|(c, _)| c).collect::<Vec<_>>();

            let versions: Vec<Version> = krates.versions().load(conn)?;
            versions
                .grouped_by(&krates)
                .into_iter()
                .map(TopVersions::from_versions)
                .zip(krates)
                .zip(recent_downloads)
                .map(|((top_versions, krate), recent_downloads)| {
                    Ok(EncodableCrate::from_minimal(
                        krate,
                        Some(&top_versions),
                        None,
                        false,
                        recent_downloads,
                    ))
                })
                .collect()
        }

        let selection = (
            Crate::as_select(),
            recent_crate_downloads::downloads.nullable(),
        );

        let new_crates = crates::table
            .left_join(recent_crate_downloads::table)
            .order(crates::created_at.desc())
            .select(selection)
            .limit(10)
            .load(conn)?;
        let just_updated = crates::table
            .left_join(recent_crate_downloads::table)
            .filter(crates::updated_at.ne(crates::created_at))
            .order(crates::updated_at.desc())
            .select(selection)
            .limit(10)
            .load(conn)?;

        let mut most_downloaded_query = crates::table
            .left_join(recent_crate_downloads::table)
            .into_boxed();
        if !config.excluded_crate_names.is_empty() {
            most_downloaded_query =
                most_downloaded_query.filter(crates::name.ne_all(&config.excluded_crate_names));
        }
        let most_downloaded = most_downloaded_query
            .then_order_by(crates::downloads.desc())
            .select(selection)
            .limit(10)
            .load(conn)?;

        let mut most_recently_downloaded_query = crates::table
            .inner_join(recent_crate_downloads::table)
            .into_boxed();
        if !config.excluded_crate_names.is_empty() {
            most_recently_downloaded_query = most_recently_downloaded_query
                .filter(crates::name.ne_all(&config.excluded_crate_names));
        }
        let most_recently_downloaded = most_recently_downloaded_query
            .then_order_by(recent_crate_downloads::downloads.desc())
            .select(selection)
            .limit(10)
            .load(conn)?;

        let popular_keywords = keywords::table
            .order(keywords::crates_cnt.desc())
            .limit(10)
            .load(conn)?
            .into_iter()
            .map(Keyword::into)
            .collect::<Vec<EncodableKeyword>>();

        let popular_categories = Category::toplevel(conn, "crates", 10, 0)?
            .into_iter()
            .map(Category::into)
            .collect::<Vec<EncodableCategory>>();

        Ok(Json(json!({
            "num_downloads": num_downloads,
            "num_crates": num_crates,
            "new_crates": encode_crates(conn, new_crates)?,
            "most_downloaded": encode_crates(conn, most_downloaded)?,
            "most_recently_downloaded": encode_crates(conn, most_recently_downloaded)?,
            "just_updated": encode_crates(conn, just_updated)?,
            "popular_keywords": popular_keywords,
            "popular_categories": popular_categories,
        })))
    })
    .await
}

/// Handles the `GET /crates/:crate_id` route.
pub async fn show(app: AppState, Path(name): Path<String>, req: Parts) -> AppResult<Json<Value>> {
    spawn_blocking(move || {
        let include = req
            .query()
            .get("include")
            .map(|mode| ShowIncludeMode::from_str(mode))
            .transpose()?
            .unwrap_or_default();

        let conn = &mut *app.db_read()?;
        let krate: Crate = Crate::by_name(&name)
            .first(conn)
            .optional()?
            .ok_or_else(|| crate_not_found(&name))?;

        let versions_publishers_and_audit_actions = if include.versions {
            let mut versions_and_publishers: Vec<(Version, Option<User>)> = krate
                .all_versions()
                .left_outer_join(users::table)
                .select((versions::all_columns, users::all_columns.nullable()))
                .load(conn)?;
            versions_and_publishers.sort_by_cached_key(|(version, _)| {
                Reverse(semver::Version::parse(&version.num).ok())
            });

            let versions = versions_and_publishers
                .iter()
                .map(|(v, _)| v)
                .cloned()
                .collect::<Vec<_>>();
            Some(
                versions_and_publishers
                    .into_iter()
                    .zip(VersionOwnerAction::for_versions(conn, &versions)?)
                    .map(|((v, pb), aas)| (v, pb, aas))
                    .collect::<Vec<_>>(),
            )
        } else {
            None
        };
        let ids = versions_publishers_and_audit_actions
            .as_ref()
            .map(|vps| vps.iter().map(|v| v.0.id).collect());

        let kws = if include.keywords {
            Some(
                CrateKeyword::belonging_to(&krate)
                    .inner_join(keywords::table)
                    .select(keywords::all_columns)
                    .load(conn)?,
            )
        } else {
            None
        };
        let cats = if include.categories {
            Some(
                CrateCategory::belonging_to(&krate)
                    .inner_join(categories::table)
                    .select(categories::all_columns)
                    .load(conn)?,
            )
        } else {
            None
        };
        let recent_downloads = if include.downloads {
            RecentCrateDownloads::belonging_to(&krate)
                .select(recent_crate_downloads::downloads)
                .get_result(conn)
                .optional()?
        } else {
            None
        };

        let badges = if include.badges { Some(vec![]) } else { None };

        let top_versions = if include.versions {
            Some(krate.top_versions(conn)?)
        } else {
            None
        };

        let encodable_crate = EncodableCrate::from(
            krate.clone(),
            top_versions.as_ref(),
            ids,
            kws.as_deref(),
            cats.as_deref(),
            badges,
            false,
            recent_downloads,
        );
        let encodable_versions = versions_publishers_and_audit_actions.map(|vpa| {
            vpa.into_iter()
                .map(|(v, pb, aas)| EncodableVersion::from(v, &krate.name, pb, aas))
                .collect::<Vec<_>>()
        });
        let encodable_keywords = kws.map(|kws| {
            kws.into_iter()
                .map(Keyword::into)
                .collect::<Vec<EncodableKeyword>>()
        });
        let encodable_cats = cats.map(|cats| {
            cats.into_iter()
                .map(Category::into)
                .collect::<Vec<EncodableCategory>>()
        });
        Ok(Json(json!({
            "crate": encodable_crate,
            "versions": encodable_versions,
            "keywords": encodable_keywords,
            "categories": encodable_cats,
        })))
    })
    .await
}

#[derive(Debug)]
struct ShowIncludeMode {
    versions: bool,
    keywords: bool,
    categories: bool,
    badges: bool,
    downloads: bool,
}

impl Default for ShowIncludeMode {
    fn default() -> Self {
        // Send everything for legacy clients that expect the full response
        Self {
            versions: true,
            keywords: true,
            categories: true,
            badges: true,
            downloads: true,
        }
    }
}

impl ShowIncludeMode {
    const INVALID_COMPONENT: &'static str =
        "invalid component for ?include= (expected 'versions', 'keywords', 'categories', 'badges', 'downloads', or 'full')";
}

impl FromStr for ShowIncludeMode {
    type Err = BoxedAppError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut mode = Self {
            versions: false,
            keywords: false,
            categories: false,
            badges: false,
            downloads: false,
        };
        for component in s.split(',') {
            match component {
                "" => {}
                "full" => {
                    mode = Self {
                        versions: true,
                        keywords: true,
                        categories: true,
                        badges: true,
                        downloads: true,
                    }
                }
                "versions" => mode.versions = true,
                "keywords" => mode.keywords = true,
                "categories" => mode.categories = true,
                "badges" => mode.badges = true,
                "downloads" => mode.downloads = true,
                _ => return Err(bad_request(Self::INVALID_COMPONENT)),
            }
        }
        Ok(mode)
    }
}

/// Handles the `GET /crates/:crate_id/:version/readme` route.
pub async fn readme(
    app: AppState,
    Path((crate_name, version)): Path<(String, String)>,
    req: Parts,
) -> Response {
    let redirect_url = app.storage.readme_location(&crate_name, &version);
    if req.wants_json() {
        Json(json!({ "url": redirect_url })).into_response()
    } else {
        redirect(redirect_url)
    }
}

/// Handles the `GET /crates/:crate_id/versions` route.
// FIXME: Not sure why this is necessary since /crates/:crate_id returns
// this information already, but ember is definitely requesting it
pub async fn versions(state: AppState, Path(crate_name): Path<String>) -> AppResult<Json<Value>> {
    spawn_blocking(move || {
        let conn = &mut *state.db_read()?;

        let krate: Crate = Crate::by_name(&crate_name)
            .first(conn)
            .optional()?
            .ok_or_else(|| crate_not_found(&crate_name))?;

        let mut versions_and_publishers: Vec<(Version, Option<User>)> = krate
            .all_versions()
            .left_outer_join(users::table)
            .select((versions::all_columns, users::all_columns.nullable()))
            .load(conn)?;

        versions_and_publishers
            .sort_by_cached_key(|(version, _)| Reverse(semver::Version::parse(&version.num).ok()));

        let versions = versions_and_publishers
            .iter()
            .map(|(v, _)| v)
            .cloned()
            .collect::<Vec<_>>();
        let versions = versions_and_publishers
            .into_iter()
            .zip(VersionOwnerAction::for_versions(conn, &versions)?)
            .map(|((v, pb), aas)| EncodableVersion::from(v, &crate_name, pb, aas))
            .collect::<Vec<_>>();

        Ok(Json(json!({ "versions": versions })))
    })
    .await
}

/// Handles the `GET /crates/:crate_id/reverse_dependencies` route.
pub async fn reverse_dependencies(
    app: AppState,
    Path(name): Path<String>,
    req: Parts,
) -> AppResult<Json<Value>> {
    spawn_blocking(move || {
        let pagination_options = PaginationOptions::builder().gather(&req)?;
        let conn = &mut *app.db_read()?;

        let krate: Crate = Crate::by_name(&name)
            .first(conn)
            .optional()?
            .ok_or_else(|| crate_not_found(&name))?;

        let (rev_deps, total) = krate.reverse_dependencies(conn, pagination_options)?;
        let rev_deps: Vec<_> = rev_deps
            .into_iter()
            .map(|dep| EncodableDependency::from_reverse_dep(dep, &krate.name))
            .collect();

        let version_ids: Vec<i32> = rev_deps.iter().map(|dep| dep.version_id).collect();

        let versions_and_publishers: Vec<(Version, String, Option<User>)> = versions::table
            .filter(versions::id.eq_any(version_ids))
            .inner_join(crates::table)
            .left_outer_join(users::table)
            .select((
                versions::all_columns,
                crates::name,
                users::all_columns.nullable(),
            ))
            .load(conn)?;
        let versions = versions_and_publishers
            .iter()
            .map(|(v, _, _)| v)
            .cloned()
            .collect::<Vec<_>>();
        let versions = versions_and_publishers
            .into_iter()
            .zip(VersionOwnerAction::for_versions(conn, &versions)?)
            .map(|((version, krate_name, published_by), actions)| {
                EncodableVersion::from(version, &krate_name, published_by, actions)
            })
            .collect::<Vec<_>>();

        Ok(Json(json!({
            "dependencies": rev_deps,
            "versions": versions,
            "meta": { "total": total },
        })))
    })
    .await
}
