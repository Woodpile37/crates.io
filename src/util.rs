use std::cmp;

pub use self::bytes_request::BytesRequest;
pub use self::io_util::{read_fill, read_le_u32};
pub use self::request_helpers::*;

mod bytes_request;
pub mod errors;
mod io_util;
mod request_helpers;
pub mod rfc3339;
pub mod token;
pub mod tracing;

#[derive(Debug, Copy, Clone)]
pub struct Maximums {
    pub max_upload_size: u64,
    pub max_unpack_size: u64,
}

impl Maximums {
    pub fn new(
        krate_max_upload: Option<i32>,
        app_max_upload: u64,
        app_max_unpack: u64,
    ) -> Maximums {
        let max_upload_size = krate_max_upload.map(|m| m as u64).unwrap_or(app_max_upload);
        let max_unpack_size = cmp::max(app_max_unpack, max_upload_size);
        Maximums {
            max_upload_size,
            max_unpack_size,
        }
    }
}
