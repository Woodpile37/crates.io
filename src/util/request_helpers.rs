use http::header::AsHeaderName;
use http::HeaderMap;

pub trait HeaderMapExt {
    /// Returns the value of the request header, or an empty slice if it is not
    /// present.
    ///
    /// If a header appears multiple times, this will return only one of them.
    ///
    /// If the header value is invalid utf8, an empty slice will be returned.
    fn get_str_or_default<K: AsHeaderName>(&self, key: K) -> &str;
}

impl HeaderMapExt for HeaderMap {
    fn get_str_or_default<K: AsHeaderName>(&self, key: K) -> &str {
        self.get(key)
            .map(|value| value.to_str().unwrap_or_default())
            .unwrap_or_default()
    }
}
