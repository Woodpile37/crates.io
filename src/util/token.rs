use diesel::{deserialize::FromSql, pg::Pg, serialize::ToSql, sql_types::Bytea};
use rand::{distributions::Uniform, rngs::OsRng, Rng};
use secrecy::{ExposeSecret, SecretString, SecretVec};
use sha2::{Digest, Sha256};

const TOKEN_LENGTH: usize = 32;

/// NEVER CHANGE THE PREFIX OF EXISTING TOKENS!!! Doing so will implicitly
/// revoke all the tokens, disrupting production users.
const TOKEN_PREFIX: &str = "cio";

#[derive(FromSqlRow, AsExpression)]
#[diesel(sql_type = Bytea)]
pub struct HashedToken(SecretVec<u8>);

impl HashedToken {
    pub(crate) fn parse(plaintext: &str) -> Option<Self> {
        // This will both reject tokens without a prefix and tokens of the wrong kind.
        if !plaintext.starts_with(TOKEN_PREFIX) {
            return None;
        }

        let sha256 = Self::hash(plaintext).into();
        Some(Self(sha256))
    }

    pub fn hash(plaintext: &str) -> Vec<u8> {
        Sha256::digest(plaintext.as_bytes()).as_slice().to_vec()
    }
}

impl std::fmt::Debug for HashedToken {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str("HashedToken")
    }
}

impl ToSql<Bytea, Pg> for HashedToken {
    fn to_sql(&self, out: &mut diesel::serialize::Output<'_, '_, Pg>) -> diesel::serialize::Result {
        ToSql::<Bytea, Pg>::to_sql(&self.0.expose_secret(), &mut out.reborrow())
    }
}

impl FromSql<Bytea, Pg> for HashedToken {
    fn from_sql(bytes: diesel::pg::PgValue<'_>) -> diesel::deserialize::Result<Self> {
        let bytes: Vec<u8> = FromSql::<Bytea, Pg>::from_sql(bytes)?;
        Ok(Self(bytes.into()))
    }
}

#[derive(Debug)]
pub struct PlainToken(SecretString);

impl PlainToken {
    pub(crate) fn generate() -> Self {
        let plaintext = format!(
            "{}{}",
            TOKEN_PREFIX,
            generate_secure_alphanumeric_string(TOKEN_LENGTH)
        )
        .into();

        Self(plaintext)
    }

    pub fn hashed(&self) -> HashedToken {
        let sha256 = HashedToken::hash(self.expose_secret()).into();
        HashedToken(sha256)
    }
}

impl ExposeSecret<String> for PlainToken {
    fn expose_secret(&self) -> &String {
        self.0.expose_secret()
    }
}

fn generate_secure_alphanumeric_string(len: usize) -> String {
    const CHARS: &[u8] = b"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";

    OsRng
        .sample_iter(Uniform::from(0..CHARS.len()))
        .map(|idx| CHARS[idx] as char)
        .take(len)
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use googletest::prelude::*;

    #[test]
    fn test_generated_and_parse() {
        let token = PlainToken::generate();
        assert_that!(token.expose_secret(), starts_with(TOKEN_PREFIX));
        assert_eq!(
            token.hashed().0.expose_secret(),
            Sha256::digest(token.expose_secret().as_bytes()).as_slice()
        );

        let parsed =
            HashedToken::parse(token.expose_secret()).expect("failed to parse back the token");
        assert_eq!(parsed.0.expose_secret(), token.hashed().0.expose_secret());
    }

    #[test]
    fn test_parse_no_kind() {
        assert!(HashedToken::parse("nokind").is_none());
    }
}
