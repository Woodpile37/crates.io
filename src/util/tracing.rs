use sentry::integrations::tracing::EventFilter;
use tracing::Level;
use tracing::Metadata;
use tracing_subscriber::filter::LevelFilter;
use tracing_subscriber::{prelude::*, EnvFilter};

/// Initializes the `tracing` logging framework.
///
/// Regular CLI output is influenced by the optional
/// [`RUST_LOG`](tracing_subscriber::filter::EnvFilter) environment variable
/// and is showing all `INFO` level events by default.
///
/// This function also sets up the Sentry error reporting integration for the
/// `tracing` framework, which is hardcoded to include all `INFO` level events.
pub fn init() {
    init_with_default_level(LevelFilter::INFO)
}

fn init_with_default_level(level: LevelFilter) {
    let env_filter = EnvFilter::builder()
        .with_default_directive(level.into())
        .from_env_lossy();

    let log_layer = tracing_subscriber::fmt::layer()
        .compact()
        .without_time()
        .with_filter(env_filter);

    let sentry_layer = sentry::integrations::tracing::layer()
        .event_filter(event_filter)
        .with_filter(LevelFilter::INFO);

    tracing_subscriber::registry()
        .with(log_layer)
        .with(sentry_layer)
        .init();
}

pub fn event_filter(metadata: &Metadata<'_>) -> EventFilter {
    match metadata.level() {
        &Level::ERROR if metadata.target() == "http" => EventFilter::Breadcrumb,
        &Level::ERROR => EventFilter::Exception,
        &Level::WARN | &Level::INFO => EventFilter::Breadcrumb,
        &Level::DEBUG | &Level::TRACE => EventFilter::Ignore,
    }
}

/// Initializes the `tracing` logging framework for usage in tests.
pub fn init_for_test() {
    let env_filter = EnvFilter::builder()
        .with_default_directive(LevelFilter::INFO.into())
        .from_env_lossy();

    let _ = tracing_subscriber::fmt()
        .compact()
        .with_env_filter(env_filter)
        .without_time()
        .with_test_writer()
        .try_init();
}
