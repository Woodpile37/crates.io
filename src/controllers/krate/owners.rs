//! All routes related to managing owners of a crate

use crate::auth::AuthCheck;
use crate::controllers::prelude::*;
use crate::models::token::EndpointScope;
use crate::models::{Crate, Owner, Rights, Team, User};
use crate::util::errors::crate_not_found;
use crate::views::EncodableOwner;
use axum::body::Bytes;
use tokio::runtime::Handle;

/// Handles the `GET /crates/:crate_id/owners` route.
pub async fn owners(state: AppState, Path(crate_name): Path<String>) -> AppResult<Json<Value>> {
    spawn_blocking(move || {
        let conn = &mut *state.db_read()?;
        let krate: Crate = Crate::by_name(&crate_name)
            .first(conn)
            .optional()?
            .ok_or_else(|| crate_not_found(&crate_name))?;

        let owners = krate
            .owners(conn)?
            .into_iter()
            .map(Owner::into)
            .collect::<Vec<EncodableOwner>>();

        Ok(Json(json!({ "users": owners })))
    })
    .await
}

/// Handles the `GET /crates/:crate_id/owner_team` route.
pub async fn owner_team(state: AppState, Path(crate_name): Path<String>) -> AppResult<Json<Value>> {
    spawn_blocking(move || {
        let conn = &mut *state.db_read()?;
        let krate: Crate = Crate::by_name(&crate_name)
            .first(conn)
            .optional()?
            .ok_or_else(|| crate_not_found(&crate_name))?;

        let owners = Team::owning(&krate, conn)?
            .into_iter()
            .map(Owner::into)
            .collect::<Vec<EncodableOwner>>();

        Ok(Json(json!({ "teams": owners })))
    })
    .await
}

/// Handles the `GET /crates/:crate_id/owner_user` route.
pub async fn owner_user(state: AppState, Path(crate_name): Path<String>) -> AppResult<Json<Value>> {
    spawn_blocking(move || {
        let conn = &mut *state.db_read()?;
        let krate: Crate = Crate::by_name(&crate_name)
            .first(conn)
            .optional()?
            .ok_or_else(|| crate_not_found(&crate_name))?;

        let owners = User::owning(&krate, conn)?
            .into_iter()
            .map(Owner::into)
            .collect::<Vec<EncodableOwner>>();

        Ok(Json(json!({ "users": owners })))
    })
    .await
}

/// Handles the `PUT /crates/:crate_id/owners` route.
pub async fn add_owners(
    app: AppState,
    Path(crate_name): Path<String>,
    req: BytesRequest,
) -> AppResult<Json<Value>> {
    spawn_blocking(move || modify_owners(&app, &crate_name, &req, true)).await
}

/// Handles the `DELETE /crates/:crate_id/owners` route.
pub async fn remove_owners(
    app: AppState,
    Path(crate_name): Path<String>,
    req: BytesRequest,
) -> AppResult<Json<Value>> {
    spawn_blocking(move || modify_owners(&app, &crate_name, &req, false)).await
}

/// Parse the JSON request body of requests to modify the owners of a crate.
///
/// The format is:
///
/// ```json
/// {"owners": ["username", "github:org:team", ...]}
/// ```
fn parse_owners_request(req: &Request<Bytes>) -> AppResult<Vec<String>> {
    #[derive(Deserialize)]
    struct Request {
        // identical, for back-compat (owners preferred)
        users: Option<Vec<String>>,
        owners: Option<Vec<String>>,
    }
    let request: Request =
        serde_json::from_slice(req.body()).map_err(|_| cargo_err("invalid json request"))?;
    request
        .owners
        .or(request.users)
        .ok_or_else(|| cargo_err("invalid json request"))
}

fn modify_owners(
    app: &AppState,
    crate_name: &str,
    req: &Request<Bytes>,
    add: bool,
) -> AppResult<Json<Value>> {
    let logins = parse_owners_request(req)?;

    let conn = &mut *app.db_write()?;
    let auth = AuthCheck::default()
        .with_endpoint_scope(EndpointScope::ChangeOwners)
        .for_crate(crate_name)
        .check(req, conn)?;

    let user = auth.user();

    conn.transaction(|conn| {
        let krate: Crate = Crate::by_name(crate_name)
            .first(conn)
            .optional()?
            .ok_or_else(|| crate_not_found(crate_name))?;

        let owners = krate.owners(conn)?;

        match Handle::current().block_on(user.rights(app, &owners))? {
            Rights::Full => {}
            // Yes!
            Rights::Publish => {
                return Err(cargo_err(
                    "team members don't have permission to modify owners",
                ));
            }
            Rights::None => {
                return Err(cargo_err("only owners have permission to modify owners"));
            }
        }

        let comma_sep_msg = if add {
            let mut msgs = Vec::with_capacity(logins.len());
            for login in &logins {
                let login_test =
                    |owner: &Owner| owner.login().to_lowercase() == *login.to_lowercase();
                if owners.iter().any(login_test) {
                    return Err(cargo_err(format_args!("`{login}` is already an owner")));
                }
                let msg = krate.owner_add(app, conn, user, login)?;
                msgs.push(msg);
            }
            msgs.join(",")
        } else {
            for login in &logins {
                krate.owner_remove(conn, login)?;
            }
            if User::owning(&krate, conn)?.is_empty() {
                return Err(cargo_err(
                    "cannot remove all individual owners of a crate. \
                     Team member don't have permission to modify owners, so \
                     at least one individual owner is required.",
                ));
            }
            "owners successfully removed".to_owned()
        };

        Ok(Json(json!({ "ok": true, "msg": comma_sep_msg })))
    })
}
