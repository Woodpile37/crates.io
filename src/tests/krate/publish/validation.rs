use crate::builders::PublishBuilder;
use crate::util::{RequestHelper, TestApp};
use crates_io::models::krate::MAX_NAME_LENGTH;
use googletest::prelude::*;
use http::StatusCode;
use insta::assert_json_snapshot;

#[test]
fn empty_json() {
    let (app, _, _, token) = TestApp::full().with_token();

    let (_json, tarball) = PublishBuilder::new("foo", "1.0.0").build();
    let body = PublishBuilder::create_publish_body("{}", &tarball);

    let response = token.publish_crate(body);
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.json());
    assert_that!(app.stored_files(), empty());
}

#[test]
fn invalid_names() {
    let (app, _, _, token) = TestApp::full().with_token();

    let bad_name = |name: &str| {
        let crate_to_publish = PublishBuilder::new(name, "1.0.0");
        let response = token.publish_crate(crate_to_publish);
        assert_eq!(response.status(), StatusCode::OK);
        assert_json_snapshot!(response.json());
    };

    bad_name("");
    bad_name("foo bar");
    bad_name(&"a".repeat(MAX_NAME_LENGTH + 1));
    bad_name("snow☃");
    bad_name("áccênts");

    bad_name("std");
    bad_name("STD");
    bad_name("compiler-rt");
    bad_name("compiler_rt");
    bad_name("coMpiLer_Rt");

    assert_that!(app.stored_files(), empty());
}

#[test]
fn invalid_version() {
    let (app, _, _, token) = TestApp::full().with_token();

    let (json, tarball) = PublishBuilder::new("foo", "1.0.0").build();
    let new_json = json.replace(r#""vers":"1.0.0""#, r#""vers":"broken""#);
    assert_ne!(json, new_json);
    let body = PublishBuilder::create_publish_body(&new_json, &tarball);

    let response = token.publish_crate(body);
    assert_json_snapshot!(response.json());
    assert_that!(app.stored_files(), empty());
}

#[test]
fn license_and_description_required() {
    let (app, _, _, token) = TestApp::full().with_token();

    let crate_to_publish = PublishBuilder::new("foo_metadata", "1.1.0")
        .unset_license()
        .unset_description();

    let response = token.publish_crate(crate_to_publish);
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.json());

    let crate_to_publish = PublishBuilder::new("foo_metadata", "1.1.0").unset_description();

    let response = token.publish_crate(crate_to_publish);
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.json());

    let crate_to_publish = PublishBuilder::new("foo_metadata", "1.1.0")
        .unset_license()
        .license_file("foo")
        .unset_description();

    let response = token.publish_crate(crate_to_publish);
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.json());

    assert_that!(app.stored_files(), empty());
}

#[test]
fn invalid_license() {
    let (app, _, _, token) = TestApp::full().with_token();

    let response =
        token.publish_crate(PublishBuilder::new("foo", "1.0.0").license("MIT AND foobar"));
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.json());
    assert_that!(app.stored_files(), empty());
}

#[test]
fn invalid_urls() {
    let (app, _, _, token) = TestApp::full().with_token();

    let response = token.publish_crate(
        PublishBuilder::new("foo", "1.0.0").documentation("javascript:alert('boom')"),
    );
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.json());

    assert_that!(app.stored_files(), empty());
}
