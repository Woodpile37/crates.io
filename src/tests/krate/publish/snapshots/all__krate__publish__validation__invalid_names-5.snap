---
source: src/tests/krate/publish/validation.rs
expression: response.into_json()
---
{
  "errors": [
    {
      "detail": "invalid character `á` in crate name: `áccênts`, the first character must be an ASCII character"
    }
  ]
}
